### PR TITLE
feat!: handle jwks verification agnostically

### PR DESCRIPTION
**Pull request by toomuchdesign**
This PR is an initial attempt to bring together the proof of concept proposed by @gnarea in #224 and a set of integration tests running against:

- auth0
- a mocked auth0 server

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -39,7 +39,7 @@
   "scripts": {
     "test": "jest ./test/test.js --coverage",
     "test:ci": "npm run lint && npm run test -- --ci --coverageReporters=json && npm run test:types",
-    "test:integration": "jest ./test/integration.test.js",
+    "test:integration": "jest ./test/*.integration.test.js",
```
Comment (simoneb) [inline package.json:42]: let's move them to dedicated folders so we don't have to do this trick. let's have a unit and integration folders

**Review by simoneb — COMMENTED** [review]
ok but wait a second, if we do this change, the API will break because we can no longer assume that we're hitting auth0 right?

**Comment by toomuchdesign** [comment]
Yes, absolutely, this refactor would imply a quite radical breaking change since `domain` is going to be an absolute url pointing to the relevant endpoint returning the `jwks` keys.

In other words: we would remove the `domain` options to rename it to something else.

BTW, this is a work in progress PR to check whether there are other blocker on our way. I expect further changes to the API.

**Comment by simoneb** [comment]
fair enough, thanks for clarifying. in that case let's be aware that before we can merge this change we'll need to make a decision as we discussed elsewhere, either fork the repo or rename this one, after deprecating the current package name

**Comment by toomuchdesign** [comment]
Tests are all green beside the integration against local mocked server on Node14. Could it be caused by an unexpected behaviour of the mocked sever on such environment?

**Comment by toomuchdesign** [comment]
Wondering whether/how we should aim to replace existing vendor-specific oauth plugins like:

- https://github.com/francisbrito/fastify-aws-cognito
- https://github.com/flexper/fastify-auth-prisma

..and whether [fastify-oauth2](https://github.com/fastify/fastify-oauth2) provides what we are doing here.

**Comment by simoneb** [comment]
> ..and whether [fastify-oauth2](https://github.com/fastify/fastify-oauth2) provides what we are doing here.

it may, yes, worth checking, also worth checking why auth0 is not in the list of presets, as it may shed some light on whether that plugin actually supersedes this one

**Comment by toomuchdesign** [comment]
[fastify-oauth2](https://github.com/fastify/fastify-oauth2) seems to only focus on granting user authentication, while `fastify-auth0-verify` does a bearer authentication check based on a [JSON Web Key Set](https://auth0.com/docs/secure/tokens/json-web-tokens/json-web-key-sets).

Not sure yet whether such functionality should be shipped as a standalone plugin, but this the specific domain we currently dealing with.

**Comment by toomuchdesign** [comment]
So, I renamed `domain` option to `jwksUrl`. We need to check whether the other existing options should be reviewed/renamed to make sense out of auth0 context.

**Comment by toomuchdesign** [comment]
The PR is ready to be reviewed and tested against possibly existing services. I left the existing options untouched since the seem to make sense even in contexts different then auth0.

The only thing left on the list consists of renaming the existing internal `auth0` fastify decorators to the new name this library is going to take.

@gnarea happy to integrate your feedback

**Comment by simoneb** [comment]
This change so far makes sense to me, thanks @toomuchdesign .

@gnarea happy to hear your feedback before proceeding further, thanks

**Comment by gnarea** [comment]
Hey folks. 👋🏾 Once again, apologies for the tardy reply. And **thank you so much** for all the effort you put into this change!

The PR looks good to me, and [I've just tried it in our app](https://github.com/relaycorp/veraid-authority/pull/153) and it all works perfectly.

By the way, we're using [`navikt/mock-oauth2-server`](https://github.com/navikt/mock-oauth2-server) in that app, which is different from the mock server you're using. So no interoperability issues there either. 🎉

**Comment by gnarea** [comment]
Hey @simoneb 👋🏾 

I haven't thought of that third option, but I think it makes a lot of sense and it may be worth checking with the folks behind fastify-jwt.

Failing that, we could keep this repo, but rename it and create a new NPM package. I think the key words for the new repo/package are "Fastify" and "JWKS", so something like `fastify-jwks-verify` or simply `fastify-jwks` could work.

Wdyt?

**Comment by simoneb** [comment]
I asked fastify's team input on this here https://github.com/fastify/help/issues/873

**Comment by simoneb** [comment]
I like Matteo's comment in that conversation, to create a new generic package but preserve this one, which will end up being a very thin, auth0-specific wrapper, around the new package.

Let's proceed in that direction then:

- fork this repo and apply the changes in this PR
- release that repo as a new package
- keep this repo but change the implementation to use the generic package

**Comment by simoneb** [comment]
Closing this PR as it's been addressed by creating https://github.com/nearform/fastify-jwt-jwks. See more in #224

**Pull request closed without merging**